### PR TITLE
offset, len, and stride in words not bytes (or 32b)

### PR DESCRIPTION
**Pull request by makslevental**
The PR changes `offset` and `len` on `dma_bd` to be denominated in **multiple of the element type width**, e.g., `memref<32xi16>` with `offset=4` and `len=8` will send **elements** 4 through 12 inclusive (3rd through 11th). 

The arithmetic to convert between the element type width and 32b alignments is now done in the op definition (~cf. `getLenIn32bWords` and `getOffsetIn32bWords`~ cf. `getLenInBytes` and `getOffsetInBytes` [the spec is in bits but libxaie is in bytes :shrug:]). 

Note, this PR does **not** update [AIEDmaToIpu::IpuDmaMemcpyNdOp](https://github.com/Xilinx/mlir-aie/blob/174398da1e140bcd9b4527aa0697ae3ec8c50e1d/lib/Dialect/AIEX/Transforms/AIEDmaToIpu.cpp#L249) because if I understand correctly, it's already treated in this way there (though I could be wrong - I am not familiar with that op).

### CODE REVIEW

File: include/aie/Dialect/AIE/IR/AIEOps.td
Comment (makslevental) [reply include/aie/Dialect/AIE/IR/AIEOps.td:762]: hazy morning... thanks for the spotting

**Comment by jgmelber** [comment]
Adding @AndraBisca here because this likely impacts the object FIFO join/distribute

**Comment by AndraBisca** [comment]
> Adding @AndraBisca here because this likely impacts the object FIFO join/distribute

Yes that will impact some of the calculations for the join and distribute, as well as the generation of AIE_DMABdOps in the objectFifo lowering. I think mainly [this ](https://github.com/Xilinx/mlir-aie/blob/main/lib/Dialect/AIE/Transforms/AIEObjectFifoStatefulTransform.cpp#L208)function needs to be updated to reflect the change in this PR. I'll take a look tomorrow and confirm.

File: lib/Targets/AIETargetAirbin.cpp
Comment (jgmelber) [inline lib/Targets/AIETargetAirbin.cpp:84]: This should still be 16 since this target only considers AIE not AIE-ML. Better yet should be using the Device Models.

**Comment by jgmelber** [comment]
Adding a note on [AIEDmaToIpu::IpuDmaMemcpyNdOp](https://github.com/Xilinx/mlir-aie/blob/174398da1e140bcd9b4527aa0697ae3ec8c50e1d/lib/Dialect/AIEX/Transforms/AIEDmaToIpu.cpp#L249), the hardware wraps/strides/offsets are all 4B aligned and the memref type is restricted to i32 or f32 for now. A future PR could incorporate these improvements to support more data types.

**Review by jgmelber — APPROVED** [review]
Good improvements! I'd wait on @AndraBisca's check to merge. It's worth adding a note that this removes the skeleton AB-mode support found in AIE (not AIE-ML) hardware.

**Comment by makslevental** [comment]
> I think mainly [this ](https://github.com/Xilinx/mlir-aie/blob/main/lib/Dialect/AIE/Transforms/AIEObjectFifoStatefulTransform.cpp#L208)function needs to be updated to reflect the change in this PR. I'll take a look tomorrow and confirm.

I already updated `AIEObjectFifoStatefulTransform` in the latest commit (actually the only place that needed to be patched was [here](https://github.com/Xilinx/mlir-aie/blob/57a29a806f43494191850c2e6339fd21fd54fd5c/lib/Dialect/AIE/Transforms/AIEObjectFifoStatefulTransform.cpp#L752)) but I'm happy to wait for a signoff from the resident expert :)

File: lib/Targets/AIETargetAirbin.cpp
Comment (makslevental) [reply lib/Targets/AIETargetAirbin.cpp:84]: I'm just gonna change it back - that thing is coupled to a bunch of `static_assert`s and statically sized arrays. cc @jnider for when you come back to airbin...

File: include/aie/Dialect/AIE/IR/AIEOps.td
Comment (stephenneuendorffer) [inline include/aie/Dialect/AIE/IR/AIEOps.td:846]: This doesn't make sense to me...shouldn't it me in multiples of the element type?

Comment (makslevental) [reply include/aie/Dialect/AIE/IR/AIEOps.td:846]: comment is wrong but impl is right (ie what you said)

Comment (makslevental) [reply include/aie/Dialect/AIE/IR/AIEOps.td:846]: But actually I realize I lied and didn't patch the n-d addressing too.

Comment (makslevental) [reply include/aie/Dialect/AIE/IR/AIEOps.td:846]: Top commit patches size and stride as well; libxaie says 

```
* @note		None. The stepsize and wrap parameters operate at 32 bit
*		granularity. 
```

but I take that to mean just stride (what they're calling stepsize) right? Because what would it even mean for size ("wrap") to be at the 32b granularity (it's the loop bound...). Sanity check here would be much appreciated @stephenneuendorffer .

File: lib/Dialect/AIE/IR/AIEDialect.cpp
Comment (makslevental) [inline lib/Dialect/AIE/IR/AIEDialect.cpp:1566]: Dunno what this is about but it's wrong for at least 2 reasons (assumes 4 bytes per element, and then for some reason computes a "byte volume"???).

Comment (stephenneuendorffer) [reply lib/Dialect/AIE/IR/AIEDialect.cpp:1566]: That seems odd... it seems like there's too many '4's getting multiplied?

**Comment by AndraBisca** [comment]
> > I think mainly [this ](https://github.com/Xilinx/mlir-aie/blob/main/lib/Dialect/AIE/Transforms/AIEObjectFifoStatefulTransform.cpp#L208)function needs to be updated to reflect the change in this PR. I'll take a look tomorrow and confirm.
> 
> I already updated `AIEObjectFifoStatefulTransform` in the latest commit (actually the only place that needed to be patched was [here](https://github.com/Xilinx/mlir-aie/blob/57a29a806f43494191850c2e6339fd21fd54fd5c/lib/Dialect/AIE/Transforms/AIEObjectFifoStatefulTransform.cpp#L752)) but I'm happy to wait for a signoff from the resident expert :)

Looks good to me! Thank you for updating Maks! :)

**Review by AndraBisca — APPROVED** [review]
Looks good to me! It's so much easier to read the offset and transfer length like this imo, and the extra checks in the AIEDialect are good. Thank you for working on this Maks!

File: include/aie/Dialect/AIE/IR/AIEOps.td
Comment (AndraBisca) [inline include/aie/Dialect/AIE/IR/AIEOps.td:845]: This changed from `multiples of 32b` (which I found straightforward) to `multiples of element width` (which I find a bit confusing). What was the reason for the change? Do we always assume that the element width is 32b?

Comment (makslevental) [reply include/aie/Dialect/AIE/IR/AIEOps.td:845]: No - the arithmetic is done so that if you write `aie.dma_bd(%buf33_0 : memref<256xi16>, 0, 32)` then 32*2B = 64 bytes will transferred.

File: include/aie/Dialect/AIE/IR/AIEOps.td
Comment (denolf) [inline include/aie/Dialect/AIE/IR/AIEOps.td:757]: is it in 32b multiples or is it in element with multiples (where element width changes depending on the type)?

**Review by denolf — CHANGES_REQUESTED** [review]

File: include/aie/Dialect/AIE/IR/AIEOps.td
Comment (AndraBisca) [reply include/aie/Dialect/AIE/IR/AIEOps.td:845]: So the length is always expressing the number of elements. Ok!

File: include/aie/Dialect/AIE/IR/AIEOps.td
Comment (denolf) [reply include/aie/Dialect/AIE/IR/AIEOps.td:757]: or on line 758, does len need to be equal to 512 in both cases?

File: reference_designs/ipu-xrt/matrix_multiplication/aie2.py
Comment (jgmelber) [inline reference_designs/ipu-xrt/matrix_multiplication/aie2.py:1]: Same comment on word size. Also noticing for all of these, the length is now wrong for the 1st dimension. It is still word_size_in // 4, but should just be "s" if the length is in multiples of data type

File: reference_designs/ipu-xrt/matrix_multiplication_column/aie2.py
Comment (jgmelber) [inline reference_designs/ipu-xrt/matrix_multiplication_column/aie2.py:1]: Same comments as others about the 1D length and word_size // 2

File: reference_designs/ipu-xrt/matrix_vector_multiplication/aie2.py
Comment (jgmelber) [inline reference_designs/ipu-xrt/matrix_vector_multiplication/aie2.py:1]: Same comments as others about the 1D length and word_size // 2

File: include/aie/Dialect/AIE/IR/AIEOps.td
Comment (makslevental) [reply include/aie/Dialect/AIE/IR/AIEOps.td:757]: it's "element width multiples" - line 758 is correct. this PR allows/enables the user to (almost) only think in terms of element width, forgetting 32b alignment except in one place - the final stride dim, which does have to be a multiple of 32b. I will update this doc string to say mention that (the last stride dim).

Comment (makslevental) [reply include/aie/Dialect/AIE/IR/AIEOps.td:757]: still waking up...

File: lib/Dialect/AIE/IR/AIEDialect.cpp
Comment (makslevental) [reply lib/Dialect/AIE/IR/AIEDialect.cpp:1566]: n-d memory! `bytes^n`!

File: reference_designs/ipu-xrt/matrix_multiplication_array/aie2.py
Comment (makslevental) [reply reference_designs/ipu-xrt/matrix_multiplication_array/aie2.py:1]: that patch was done as a "go to bed see if it works in the morning" since I don't understand those designs - it mostly worked! I will polish now.

File: reference_designs/ipu-xrt/matrix_vector_multiplication/aie2.py
Comment (denolf) [inline reference_designs/ipu-xrt/matrix_vector_multiplication/aie2.py:106]: Why are we using k_in_i32s here? Does more of this PR need to be absorbed in the objFIFO?

File: reference_designs/ipu-xrt/matrix_vector_multiplication/aie2.py
Comment (denolf) [inline reference_designs/ipu-xrt/matrix_vector_multiplication/aie2.py:107]: Similar question, why is the *2 needed here

File: reference_designs/ipu-xrt/matrix_vector_multiplication/aie2.py
Comment (denolf) [reply reference_designs/ipu-xrt/matrix_vector_multiplication/aie2.py]: I was hoping to see here k and not k_in_i32s.

**Review by denolf — APPROVED** [review]
LGTM, thanks for the improvements.

File: reference_designs/ipu-xrt/matrix_vector_multiplication/aie2.py
Comment (jgmelber) [inline reference_designs/ipu-xrt/matrix_vector_multiplication/aie2.py:107]: (k, 1)

**Review by jgmelber — APPROVED** [review]

File: reference_designs/ipu-xrt/matrix_vector_multiplication/aie2.py
Comment (makslevental) [reply reference_designs/ipu-xrt/matrix_vector_multiplication/aie2.py:107]: i'm still figuring this one out... not landing yet

File: include/aie/Dialect/AIE/IR/AIEOps.td
Comment (stephenneuendorffer) [inline include/aie/Dialect/AIE/IR/AIEOps.td:844]: I know what you're trying to say, but this seems like its saying "Exists some k s.t. stride = k * 32b"   How about we just say "Addresses must be 32bit aligned"?

File: include/aie/Dialect/AIE/IR/AIEOps.td
Comment (stephenneuendorffer) [inline include/aie/Dialect/AIE/IR/AIEOps.td:846]: rather than saying "cannot", I would say "stride_0" must be a multiple of 2.  if the element type is i8, then similarly the stride parameter must be a multiple of 4.

File: include/aie/Dialect/AIE/IR/AIEOps.td
Comment (stephenneuendorffer) [inline include/aie/Dialect/AIE/IR/AIEOps.td:853]: 'words' is too ambiguous, IMO.

File: lib/Dialect/AIE/IR/AIEDialect.cpp
Comment (stephenneuendorffer) [inline lib/Dialect/AIE/IR/AIEDialect.cpp:1552]: nit: the parameter is just a number so it's a multiple of 4.  Or, don't abbreviate Bytes to B.

File: lib/Dialect/AIE/IR/AIEDialect.cpp
Comment (stephenneuendorffer) [inline lib/Dialect/AIE/IR/AIEDialect.cpp:1561]: Factor out magic constant 15, preferably to the TargetModel

File: lib/Dialect/AIE/IR/AIEDialect.cpp
Comment (stephenneuendorffer) [inline lib/Dialect/AIE/IR/AIEDialect.cpp:1564]: And these should probably have negative tests.

File: lib/Dialect/AIE/IR/AIEDialect.cpp
Comment (stephenneuendorffer) [inline lib/Dialect/AIE/IR/AIEDialect.cpp:1586]: Again, TargetModel for the magic numbers.

File: lib/Dialect/AIE/IR/AIEDialect.cpp
Comment (stephenneuendorffer) [inline lib/Dialect/AIE/IR/AIEDialect.cpp:1621]: These changes seem suspicious to me, since they seem to preclude having memref view transformations?   It's not clear to me why this is better, I guess.

File: lib/Dialect/AIE/Transforms/AIEObjectFifoStatefulTransform.cpp
Comment (stephenneuendorffer) [inline lib/Dialect/AIE/Transforms/AIEObjectFifoStatefulTransform.cpp:752]: Was this a functional error?  Do we have a test for the fix?

File: lib/Targets/AIETargetCDODirect.cpp
Comment (stephenneuendorffer) [inline lib/Targets/AIETargetCDODirect.cpp:291]: This seems like it should be in the verifier?

File: lib/Targets/AIETargetCDODirect.cpp
Comment (stephenneuendorffer) [inline lib/Targets/AIETargetCDODirect.cpp:304]: double???

File: test/Targets/AIEGenerateXAIE/aie2_nd_DMA.mlir
Comment (stephenneuendorffer) [inline test/Targets/AIEGenerateXAIE/aie2_nd_DMA.mlir:22]: I think there's a somewhat significant loss of readability here...

**Review by stephenneuendorffer — COMMENTED** [review]
This is a tricky change :)  Thanks for the effort!

File: lib/Dialect/AIE/Transforms/AIEObjectFifoStatefulTransform.cpp
Comment (makslevental) [reply lib/Dialect/AIE/Transforms/AIEObjectFifoStatefulTransform.cpp:752]: It wasn't an error - it's just no longer correct when `offset` is in terms of # of element rather than #number of bytes

File: lib/Targets/AIETargetCDODirect.cpp
Comment (makslevental) [reply lib/Targets/AIETargetCDODirect.cpp:291]: Base address is only know after the `aie-assign-buffer-addresses` pass but sure.

File: lib/Targets/AIETargetCDODirect.cpp
Comment (makslevental) [inline lib/Targets/AIETargetCDODirect.cpp:326]: Assign defaults in the unassigned inner dims

File: include/aie/Dialect/AIE/IR/AIEOps.td
Comment (makslevental) [reply include/aie/Dialect/AIE/IR/AIEOps.td:850]: this is wrong now.

File: include/aie/Dialect/AIE/IR/AIEOps.td
Comment (makslevental) [reply include/aie/Dialect/AIE/IR/AIEOps.td:848]: This is wrong now - last stride must be 1.

File: lib/Dialect/AIE/IR/AIEDialect.cpp
Comment (makslevental) [reply lib/Dialect/AIE/IR/AIEDialect.cpp:1561]: there's actually a verifyTraint/interface thing that does this check against the target model but dma_bd doesn't implement. Let me try to implement that interface instead of this.

File: include/aie/Dialect/AIE/IR/AIEOps.td
Comment (AndraBisca) [inline include/aie/Dialect/AIE/IR/AIEOps.td:838]: I think moving offset and len to the attribute dictionary should be part of another PR, centred around syntax.

**Review by AndraBisca — CHANGES_REQUESTED** [review]

File: include/aie/Dialect/AIE/IR/AIEOps.td
Comment (makslevental) [reply include/aie/Dialect/AIE/IR/AIEOps.td:838]: Sure

**Review by AndraBisca — APPROVED** [review]

**Pull request merged**
